### PR TITLE
make ProcessException pickleable

### LINKED ISSUE #70116: multiprocessing ProcessException (and subclasses) can't be pickled/unpickled
### 🐛 Describe the bug

 Error instances that inherit from [`ProcessException`](https://github.com/pytorch/pytorch/blob/1065739781cae67f7861beaaceadb736e5e52271/torch/multiprocessing/spawn.py#L12) cannot be picked/unpickled. For example:

```python
>>> import dill
>>> from torch.multiprocessing import ProcessRaisedException
>>> e = ProcessRaisedException("Oh no!", 1, 1)
>>> dill.loads(dill.dumps(e))
```

```
Traceback (most recent call last):
    dill.loads(dill.dumps(e))
  File ".../lib/python3.9/site-packages/dill/_dill.py", line 327, in loads
    return load(file, ignore, **kwds)
  File ".../lib/python3.9/site-packages/dill/_dill.py", line 313, in load
    return Unpickler(file, ignore=ignore, **kwds).load()
  File ".../lib/python3.9/site-packages/dill/_dill.py", line 525, in load
    obj = StockUnpickler.load(self)
TypeError: __init__() missing 2 required positional arguments: 'error_index' and 'error_pid'
```

This is easy to fix and I wouldn't mind submitting a PR to do so. We'd just need to override `__reduce__` on `ProcessException`:

```python
class ProcessException(Exception):
    __slots__ = ["error_index", "error_pid"]

    def __init__(self, msg: str, error_index: int, pid: int):
        super().__init__(msg)
        self.error_index = error_index
        self.pid = pid

    def __reduce__(self):
        return (self.__class__, self.args + (self.error_index, self.pid), {})
```

### Versions

Collecting environment information...
PyTorch version: 1.10.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.7 (default, Sep  9 2021, 23:20:13)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.11.0-1023-gcp-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB

Nvidia driver version: 470.86
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy==0.910
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] torch==1.10.1+cu113
[conda] Could not collect

cc @VitalyFedyunin

### PR DESCRIPTION
**Pull request by epwalsh**
Fixes #70116

Happy to add tests if you let me know the best place to put them.

cc @VitalyFedyunin

### CODE REVIEW

File: torch/multiprocessing/spawn.py [hunk 1]
```diff
@@ -17,6 +17,9 @@ def __init__(self, msg: str, error_index: int, pid: int):
         self.error_index = error_index
         self.pid = pid
 
+    def __reduce__(self):
+        return (self.__class__, self.args + (self.error_index, self.pid) + {})
+
```
Comment (ejguan) [inline torch/multiprocessing/spawn.py:22]: You can do the following to achieve such behavior. Using `args` is too hacky for me.
```py
    def __reduce__(self):
        return type(self), (self.msg, self.error_index, self.pid)
```

**Review by ejguan — COMMENTED** [review]
Could you please add tests here?
https://github.com/pytorch/pytorch/blob/00a871c5c9f4d616bb8e528bbed35e3f4d6293d8/test/test_multiprocessing_spawn.py

File: torch/multiprocessing/spawn.py (hunk 1 above)
Comment (epwalsh) [reply torch/multiprocessing/spawn.py]: There is no `msg` attribute, but I did refactor to make it more readable.

Comment (ejguan) [reply torch/multiprocessing/spawn.py]: What you are doing now is assuming there is key value `msg` in `self.args`. I prefer attaching `msg` to each Error instance.

**Review by ejguan — APPROVED** [review]
LGTM, thanks

**Pull request closed without merging**